Kernel/Net/RTL8168NetworkAdapter.cpp

```diff
@@ -0,0 +1,385 @@
+        m_rx_buffers_regions.append(region.release_nonnull());
+
+        descriptor.buffer_size = PACKET_BUFFER_SIZE;
+        descriptor.flags = RXDescriptor::Ownership; // let the NIC know it can use this descriptor
+        auto physical_address = m_rx_buffers_regions[i].physical_page(0)->paddr().get();
+        descriptor.buffer_address_low = physical_address & 0xFFFFFFFF;
+        descriptor.buffer_address_high = (u64)physical_address >> 32; // cast to prevent shift count >= with of type warnings in 32 bit systems
+    }
+    rx_descriptors[number_of_rx_descriptors - 1].flags = rx_descriptors[number_of_rx_descriptors - 1].flags | RXDescriptor::EndOfRing;
+}
+
+UNMAP_AFTER_INIT void RTL8168NetworkAdapter::initialize_tx_descriptors()
+{
+    auto* tx_descriptors = (TXDescriptor*)m_tx_descriptors_region->vaddr().as_ptr();
+    for (size_t i = 0; i < number_of_tx_descriptors; ++i) {
+        auto& descriptor = tx_descriptors[i];
+        auto region = MM.allocate_contiguous_kernel_region(page_round_up(PACKET_BUFFER_SIZE), "RTL8168 TX buffer", Region::Access::Read | Region::Access::Write);
+        VERIFY(region);
+        m_tx_buffers_regions.append(region.release_nonnull());
+        descriptor.flags = TXDescriptor::FirstSegment | TXDescriptor::LastSegment; // FIXME: can these be omitted?
+        auto physical_address = m_tx_buffers_regions[i].physical_page(0)->paddr().get();
+        descriptor.buffer_address_low = physical_address & 0xFFFFFFFF;
+        descriptor.buffer_address_high = (u64)physical_address >> 32;
+    }
+    tx_descriptors[number_of_tx_descriptors - 1].flags = tx_descriptors[number_of_tx_descriptors - 1].flags | TXDescriptor::EndOfRing;
+}
+
+UNMAP_AFTER_INIT RTL8168NetworkAdapter::~RTL8168NetworkAdapter()
+{
+}
+
+void RTL8168NetworkAdapter::handle_irq(const RegisterState&)
+{
+    for (;;) {
+        int status = in16(REG_ISR);
+        out16(REG_ISR, status);
+
+        m_entropy_source.add_random_event(status);
+
+        dbgln_if(RTL8168_DEBUG, "RTL8168: handle_irq status={:#04x}", status);
+
+        if ((status & (INT_RXOK | INT_RXERR | INT_TXOK | INT_TXERR | INT_RX_DESCRIPTOR_UNAVAILABLE | INT_LINK_CHANGE | INT_RX_FIFO_OVERFLOW)) == 0)
+            break;
+
+        if (status & INT_RXOK) {
+            dbgln_if(RTL8168_DEBUG, "RTL8168: RX ready");
+            receive();
+        }
+        if (status & INT_RXERR) {
+            dbgln_if(RTL8168_DEBUG, "RTL8168: RX error - invalid packet");
+        }
+        if (status & INT_TXOK) {
+            dbgln_if(RTL8168_DEBUG, "RTL8168: TX complete");
+            m_wait_queue.wake_one();
+        }
+        if (status & INT_TXERR) {
+            dbgln_if(RTL8168_DEBUG, "RTL8168: TX error - invalid packet");
+        }
+        if (status & INT_RX_DESCRIPTOR_UNAVAILABLE) {
+            dmesgln("RTL8168: RX descriptor unavailable (packet lost)");
+        }
+        if (status & INT_LINK_CHANGE) {
+            m_link_up = (in8(REG_PHYSTATUS) & PHY_LINK_STATUS) == 1;
```

Comment:
I logged REG_PHYSTATUS when this IRQ came in. It came in as `0xf3`, which according to the data sheet means the link is up.
Looking at it, this is a typo: 0xF3 & 0x02 == 2, not 1 :^)

Reply:
oops x2 :sweat_smile: